### DISCUSSION TITLE
Handling Dynamic Routing in Nuxt.js

### DISCUSSION THREAD

**Discussion by sync42johnny** [nuxt/nuxt#21447] [Questions]
Hello everyone,

I've been working on a Nuxt.js project that involves dynamic routing, and I'm encountering some complexities that I hope to discuss and find solutions for with this community.

**Project Structure:**

The project involves a car-related application, and the URL structure can be quite flexible, reflecting different layers of categorization and providing specific data about cars. Here's a simplified version of the routes:

- /cars/[category]/[[subcategory]/[[provider]]/index.vue
    - /cars/sedans
    - /cars/sedans/bmw
    - /cars/sedans/electric
    - /cars/sedans/electric/tesla
    - /cars/sedans/hybrid/electric/tesla

In these routes, both the subcategory and provider parameters are optional. The parameters can be any string, and their role (as a category, subcategory, or provider) is determined by their position in the URL.

**Challenge:**

The challenge I'm facing is ensuring that the routing system correctly identifies each parameter as a category, subcategory, or provider, even when some parameters are missing.

For instance, with the path /cars/sedans/bmw, the router params return:
`{ "category": "sedans", "subcategory": "bmw", "provider": "" }`

**Discussion:**

I'm curious about how others have approached similar situations in their own Nuxt.js projects. How can I ensure that the routing system correctly identifies these parameters based on their position in the URL?

**Comment by sync42johnny** [comment]
One strategy I've attempted to use to resolve this issue involved the use of the nuxt-link component with name and params. Here's an example:

```
<nuxt-link
  :to="{
    name: 'category-subcategory-provider',
    params: { category: 'sedans', subcategory: 'electric', provider: 'tesla' }
  }"
>
  Click me
</nuxt-link>
```

This link was placed in the /pages/cars.vue file. However, it led to an error with the message: "Unhandled error during execution of setup function".
`Uncaught (in promise) Error: No match for
 {"name":"category-subcategory-provider","params":{"category":"sedans","subcategory":"electric","provider":"tesla"}}
`

**Comment by maou-shonen** [comment]
You cannot use "[[]]".
The most common real-world examples are using "-" or "_" for filling (empty strings often cause issues).

**Comment by sync42johnny** [comment]
```
import { defineNuxtModule, createResolver } from '@nuxt/kit';

export default defineNuxtModule({
  setup(options, nuxt) {
    const { resolve } = createResolver(import.meta.url);
  
    nuxt.hook('pages:extend', (pages) => {

      pages.push({
        name: 'games-category-subcategory-provider',
        path: '/games/:category/:subcategory?/:provider?',
        file: '~/pages/games/[category].vue'
      });
    });
  }
});
```


```
<nuxt-link
  :to="{
    name: `games-category-subcategory-provider___${$i18n.locale}`,
    params: { category: 'sedans', subcategory: '', provider: 'tesla' }
  }"
>
```

After navigating to the route via the nuxt-link, the params object correctly reads as { category: 'sedans', subcategory: '', provider: 'tesla' }. However, after refreshing the page, the params object changes to { category: 'sedans', subcategory: 'tesla', provider: '' }.

It seems like on refresh, the provider parameter is incorrectly being interpreted as the subcategory. The issue appears to be that after a page refresh, the optional parameters in the route definition /games/:category/:subcategory?/:provider? are getting filled in order, without regard for the parameter names.

I'm looking for a solution that ensures the 'provider' remains recognized as the 'provider' even after a page refresh, even when the route is /cars/sedans/tesla, where 'tesla' is actually a provider. This is important because my application needs to know that 'tesla' is a provider, not a subcategory. Any suggestions on how to handle this would be greatly appreciated.
